### PR TITLE
Replace AggregatorTestCase#search with AggregatorTestCase#searchAndReduce

### PR DESCRIPTION
**Pull request by jimczi**
This commit removes the ability to test the top level result of an aggregator
before it runs the final reduce. All aggregator tests that use AggregatorTestCase#search
are rewritten with AggregatorTestCase#searchAndReduce in order to ensure that we test
the final output (the one sent to the end user) rather than an intermediary result
that could be different.
This change also removes spurious commits triggered on top of a random index writer.
These commits slow down the tests and are redundant with the commits that the
random index writer performs.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalVariableWidthHistogram.java [hunk 1]
```diff
@@ -451,7 +450,8 @@ private void mergeBucketsWithPlan(List<Bucket> buckets, List<BucketRange> plan,
             }
             toMerge.add(buckets.get(startIdx)); // Don't remove the startIdx bucket because it will be replaced by the merged bucket
 
-            reduceContext.consumeBucketsAndMaybeBreak(- (toMerge.size() - 1));
+            int toRemove = toMerge.stream().mapToInt(b -> countInnerBucket(b)+1).sum();
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalVariableWidthHistogram.java:453]: Is this an actual bug you found while making this change?

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/filter/FilterAggregatorTests.java [hunk 2]
```diff
@@ -104,7 +104,7 @@ public void testRandom() throws Exception {
                     response = searchAndReduce(indexSearcher, new MatchAllDocsQuery(), builder,
                         fieldType);
                 } else {
-                    response = search(indexSearcher, new MatchAllDocsQuery(), builder, fieldType);
+                    response = searchAndReduce(indexSearcher, new MatchAllDocsQuery(), builder, fieldType);
```
Comment (not-napoleon) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/filter/FilterAggregatorTests.java:107]: I think we can remove the entire `if` block here, and the `doReduce` for loop.

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregatorTests.java [hunk 3]
```diff
@@ -57,7 +57,7 @@ public void testDoubles() throws Exception {
                 new RangeFieldMapper.Range(rangeType, 1.0D, 5.0D, true, true), // bucket 0 5
                 new RangeFieldMapper.Range(rangeType, -3.1, 4.2, true, true), // bucket -5, 0
                 new RangeFieldMapper.Range(rangeType, 4.2, 13.3, true, true), // bucket 0, 5, 10
-                new RangeFieldMapper.Range(rangeType, 42.5, 49.3, true, true), // bucket 40, 45
+                new RangeFieldMapper.Range(rangeType, 22.5, 29.3, true, true), // bucket 40, 45
```
Comment (not-napoleon) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregatorTests.java:60]: let's either update the comment to match the new range, or just delete the comments.  And in the later tests as well.

**Review by not-napoleon — APPROVED** [review]
Looks good.  Please clean up that one for loop that got missed, otherwise 👍

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/filter/FilterAggregatorTests.java (hunk 2 above)
Comment (jimczi) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/filter/FilterAggregatorTests.java]: thanks, I pushed https://github.com/elastic/elasticsearch/pull/60683/commits/1dd98d16f924040dc870ab26c8feee62ad65e3ab

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregatorTests.java (hunk 3 above)
Comment (jimczi) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregatorTests.java]: ++, I pushed https://github.com/elastic/elasticsearch/pull/60683/commits/1dd98d16f924040dc870ab26c8feee62ad65e3ab

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalVariableWidthHistogram.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalVariableWidthHistogram.java:453]: A minor one, yes. The max bucket count is not accurate when buckets are auto-merged.

**Review by nik9000 — APPROVED** [review]
I think for the auto date histo and variable width histo I'll miss being able to assert on things that aren't yet finally reduced, but it is worth it.

**Pull request merged**